### ISSUE TITLE
Error: Failed to connect to socket in arch linux

### ISSUE THREAD

**Issue by Abrahamx9-R** [LGFae/swww#108]
i try to install in archlinux but when i want to run i have the error : Failed to connect to socket: Connection refused (os error 111)

**Comment by Le-Kat** [comment]
I'm also experiencing this issue (installing both swww & swww-git from the AUR, and also compiling it manually). I can't test whether this is a wayland or X11 issue (I only have wayland installed). For the time being I recommend reverting to an version 0.7.1, since that's the latest functioning version.

**Comment by Teddy-Kun** [comment]
I can confirm this as well. This happens when I use my usual scaling of 1.25 in Hyprland. Setting the scaling to 1 fixes it for me.

**Comment by Karmak23** [comment]
Here I’ve got the issue on a yoga2 pro (3600x1800 @×2) but no problem on a Surface Go2 (1920x1200 @×1.25 if memory serves me right). [Arch linux, swww git version from a month ago]
A strange thing is that it worked correctly on the yoga2 until 3 days ago, where swww-daemon started crashing right after launch. `RUST_BACKTRACE=full` seems to show nothing interesting :

```
11:11:01 [INFO] (main) Initialization succeeded! Starting main loop...
thread 'main' panicked at 'internal error: entered unreachable code: Iterators of zip_eq have different sizes: 5760000, 1440000', /home/olive/.cache/yay/swww-git/src/swww/utils/src/comp_decomp.rs:207:9
stack backtrace:
   0:     0x55559c081c80 - <unknown>
   1:     0x55559c036bee - <unknown>
   2:     0x55559c0601d4 - <unknown>
   3:     0x55559c0831bf - <unknown>
   4:     0x55559c082dbf - <unknown>
   5:     0x55559c083e4d - <unknown>
   6:     0x55559c083924 - <unknown>
   7:     0x55559c08388c - <unknown>
   8:     0x55559c083861 - <unknown>
   9:     0x55559bfeeea2 - <unknown>
  10:     0x55559c01d1cc - <unknown>
  11:     0x55559c02c1ae - <unknown>
  12:     0x55559c009a53 - <unknown>
  13:     0x55559c032e0c - <unknown>
  14:     0x7fce18c9c790 - <unknown>
  15:     0x7fce18c9c84a - __libc_start_main
  16:     0x55559bff1995 - <unknown>
  17:                0x0 - <unknown>
Abandon (core dumped)
```

**Comment by Karmak23** [comment]
Even after a full clean / recompile, the crash still occurs.

**Comment by Horus645** [comment]
@Karmak23, thiis is very interesting indeed! I failed to reproduce this in both my notebook and my desktop, and just couldn't get my head around what was going on.

First of all, the fact the problem only manifests itself in your 3600x1800 monitor explains why I couldn't reproduce it. It may also be related to #106. 

Regarding the error, it seems we are sending an image of the wrong size somewhere along the pipeline's line. I think I will change that panic to instead just print a error at runtime. That way, at the very least you guys won't have `swww-daemon` crashing on you. I can then take my time further investigating it.

**Comment by Horus645** [comment]
@Karmak23 if it's not too much trouble, can you try out PR #109 and see if the daemon stopped panicking?

**Comment by Karmak23** [comment]
> @Karmak23 if it's not too much trouble, can you try out PR #109 and see if the daemon stopped panicking?

Hey,

sorry, I tried git clone but I don't know anything about cargo. It would be easier if you merged the branch into main, I would just rebuild my AUR package `swww-git` with latest code. If the code is not intrusive, this would eventually allow other people to be informed when swww has a problem.

I correct the informations about my monitor :
```
Output eDP-1 'SDC 0x424A Unknown' (focused)
  Current mode: 3200x1800 @ 60.000 Hz
  Position: 0,0
  Scale factor: 2.000000
  Scale filter: nearest
  Subpixel hinting: unknown
  Transform: normal
  Workspace: 5
  Max render time: off
  Adaptive sync: disabled
  Available modes:
    3200x1800 @ 60.000 Hz
    3200x1800 @ 48.011 Hz
```

Strange thing is that `swww` worked for some days before starting to crash at every launch. Sadly I didn't notice any other change in my regular upgrades (I upgrade my arch machines every one or two days…).

**Comment by Karmak23** [comment]
Ho ! tried something else :
```
mv ~/.cache/swww/eDP-1 ~/.cache/swww/eDP-1.save
rm /var/run/user/1000/swww.socket
swww-daemon &  
```

then change background : it works !!
Could this be a corrupted image ? a corrupted cache ?

**Comment by Horus645** [comment]
It is almost certainly that we are saving the cache wrong. I'm yet to figure out why thought. The big problem is that I can't reproduce it in my machines. But now I have a feeling it's probably due to the scaling.

> https://doc.rust-lang.org/cargo/reference/config.html

Nah, that's just a bunch of pixels. Won't be very helpful. But thanks!

**Comment by Karmak23** [comment]
If i can remember correctly, the image that was last used to create a wallpaper is insanely small compared to the monitor resolution. It’s 1920x1097px.
Could it be related with images not having an odd number of pixels / dimension ?

**Comment by Horus645** [comment]
> Could it be related with images not having an odd number of pixels / dimension ?

Unlikely, as I don't have that problem myself.

> It would be easier if you merged the branch into main, I would just rebuild my AUR package swww-git with latest code

I've just done that. Could please check it for me? I am more and more convinced that that is, in fact, the problem

**Comment by Karmak23** [comment]
It’s running. It said :
```
18:22:03 [INFO] (main) eDP-1 monitor's buffer size (5760000) is smaller than cache's image (23040000)
```

**Comment by Horus645** [comment]
I don't think it will crash anymore. If will generate that log every time it tries to reload the image though.

**Comment by Karmak23** [comment]
> I don't think it will crash anymore. It will generate that log every time it tries to reload the cached image though (so, only during initialization).

I’ve been cycling images every few seconds with my kids and it didn't crash. I will update here if it does. regards and good night (おやすみ from France ;-) ).

**Comment by Karmak23** [comment]
Wallpapers have been cycling for more than 24 hours without any problem with latest main branch. I don't have the button to close this issue, but I consider it resolved. Thanks a lot．
ありがとございました!

**Comment by HelionSmoker** [comment]
Running swww 0.7.3-0 from AUR, using `dwl` I'm running into this issue:
When resizing from small to big window, i seem to get:
```
Error: "Failed to receive answer: io error: Connection reset by peer (os error 104)"
```
while when resizing from big to small:
```
Error: "Failed to connect to socket: Connection refused (os error 111)"
```

Sometimes they switch. See attacked screencast for reference

https://user-images.githubusercontent.com/82761116/234317071-d9d6e4ba-3ca3-41a1-871d-2ead209d0d68.mp4

**Comment by xHallaca** [comment]
I tried to do this: clear the cache of the program, stop it and then start it again, that worked for me:

 `
$ rm -r ~/.cache/swww
$ swww kill
$ swww init
$ swww img /path/to/any/image/
`

**Comment by Diaoul** [comment]
So I've been able to reproduce the bug from the OP by changing the scaling of my monitor:
```
❯ swww init --no-daemon
WARNING: socket file /run/user/1000/swww.socket was not deleted when the previous daemon exited
08:28:11 [INFO] (main) Initialization succeeded! Starting main loop...
08:28:11 [INFO] (main) DP-7 monitor's buffer size (14745600) is smaller than cache's image (33177600)
thread 'main' panicked at 'internal error: entered unreachable code: Iterators of zip_eq have different sizes: 14745600, 8294400', /home/antoine/.cache/paru/clone/swww/src/swww-0.7.3/utils/src/comp_decomp.rs:207:9
stack backtrace:
   0:     0x55fc6a065320 - <unknown>
   1:     0x55fc6a01526e - <unknown>
   2:     0x55fc6a040304 - <unknown>
   3:     0x55fc6a06684f - <unknown>
   4:     0x55fc6a06644f - <unknown>
   5:     0x55fc6a0674c9 - <unknown>
   6:     0x55fc6a066fa4 - <unknown>
   7:     0x55fc6a066f06 - <unknown>
   8:     0x55fc6a066ef1 - <unknown>
   9:     0x55fc69fcc182 - <unknown>
  10:     0x55fc69ffb3e4 - <unknown>
  11:     0x55fc6a00a516 - <unknown>
  12:     0x55fc69fe8d13 - <unknown>
  13:     0x55fc6a0116ec - <unknown>
  14:     0x7f924d761790 - <unknown>
  15:     0x7f924d76184a - __libc_start_main
  16:     0x55fc69fced35 - <unknown>
  17:                0x0 - <unknown>
```

Clearing `swww`'s cache seems to do the trick.

```rm -rf  ~/.cache/swww/```

Then

```
❯ swww init --no-daemon
WARNING: socket file /run/user/1000/swww.socket was not deleted when the previous daemon exited
08:33:14 [INFO] (main) Initialization succeeded! Starting main loop...
08:33:14 [INFO] (main) failed to find cached image for monitor 'eDP-1'
08:33:14 [INFO] (main) failed to find cached image for monitor 'DP-4'
08:33:14 [INFO] (main) failed to find cached image for monitor 'DP-7'
08:33:14 [INFO] (main) failed to find cached image for monitor 'DP-7'
```

**Comment by YinAqu** [comment]
> Clearing `swww`'s cache seems to do the trick.
> 
> `rm -rf ~/.cache/swww/`
I've been stuck with this error for days, clearing the cache help me out.

**Comment by YtGz** [comment]
> It is almost certainly that we are saving the cache wrong. I'm yet to figure out why thought. The big problem is that I can't reproduce it in my machines. But now I have a feeling it's probably due to the scaling.
> 
> > I still have the "perhaps corrupted" cache file (5,5Mb), tell me if you want it.
> 
> Nah, that's just a bunch of pixels. Won't be very helpful. But thanks!

I have tried to debug it on my device with a monitor of resolution **2880x1800, scaling factor 2**.
The crash occurs [here](https://github.com/Horus645/swww/blob/1112c2de3f9cec59e3713dbebf7ba2afe25fd9d0/utils/src/comp_decomp.rs#L38). The reason for this seems to be the following: `get_surface_scale_factor()` initially reports a scaling factor of `1`, but when the value is logged for the second time it reports the correct scaling factor of `2`. This discrepancy is what causes the cached image to be saved with the wrong size. I know nothing about the Wayland protocol, but maybe it is related to [this](https://mail.openjdk.org/pipermail/wakefield-dev/2022-December/000072.html)?

The change of the scale factor can be observed by using [`create_surface_with_scale_callback`](https://github.com/Horus645/swww/blob/1112c2de3f9cec59e3713dbebf7ba2afe25fd9d0/daemon/src/main.rs#L291), the callback will be triggered when the scale factor switches from `1` to `2`.

Manually setting the scaling factor to a hardcoded value [here](https://github.com/Horus645/swww/blob/1112c2de3f9cec59e3713dbebf7ba2afe25fd9d0/daemon/src/main.rs#L120) fixes the issue but is of course not a proper solution.

**Comment by Horus645** [comment]
That's very interesting @YtGz. In fact, it's great help! This should no longer be a problem in `sctk-0.17` branch, since we no longer call that code in the `daemon`. Unfortunately, we also had to nuke the cache, because it was causing a lot of problems. I now realize the problems I was having with the cache in `sctk-0.17` were also most likely due to the same cause.

Currently, I am going back to my idea to just offer a `kanshi` configuration and some scripts to help setting a wallpaper automatically when connecting a new monitor, which was the original purpose of the cache anyway.

**Comment by CarloCattano** [comment]
> mv ~/.cache/swww/eDP-1 ~/.cache/swww/eDP-1.save
> rm /var/run/user/1000/swww.socket
> swww-daemon &

This did the trick for me

**Comment by krypciak** [comment]
I managed to reproduce this bug 100% of the time (on my machine):
```bash
#!/bin/sh
pkill swww; sleep 1
swww init; sleep 2

wlr-randr --output HDMI-A-1 --mode 1920x1080
sleep 3
swww img /home/krypek/.config/wallpapers/autumn.png
wlr-randr --output HDMI-A-1 --mode 1920x1200
sleep 3
swww img /home/krypek/.config/wallpapers/autumn.png
wlr-randr --output HDMI-A-1 --mode 1920x1080

# this fixes the issue
rm -rf ~/.cache/swww; rm -f /tmp/1000-runtime-dir.cAW/swww.socket
```

**Comment by thenicebro** [comment]
> Ho ! tried something else :
> 
> ```
> mv ~/.cache/swww/eDP-1 ~/.cache/swww/eDP-1.save
> rm /var/run/user/1000/swww.socket
> swww-daemon &  
> ```
> 
> then change background : it works !! Could this be a corrupted image ? a corrupted cache ?
> 
> I still have the "perhaps corrupted" cache file (5,5Mb), tell me if you want it.

It works for me, thanks for your solution

**Comment by Horus645** [comment]
Hello everyone, this might be solved in the latest release, if someone could test it out for me I'd greatly appreciate it. (If it isn't solved, I will prioritize it -- I could't do that before because I was busy rewriting the `swww-daemon`. Sorry for the delay).

**Comment by Horus645** [comment]
I am having trouble reproducing it. Changing the resolution does not crash `swww` for me, and sending a new image succeeds.

Are you sure you moved both `swww` and `swww-daemon` to your `$PATH`? If you don't want to do that, you can try using two terminals and running `./target/release/swww-daemon` directly from the top level directory in one terminal, and calling the client in another.

**Comment by LGFae** [comment]
`swww` has changed a lot since this was opened. Furthermore, it seems overall the issue has been fixed. So I will be closing this for now. If there's still a problem, I would encourage actually making a new issue, since a long time has passed.
